### PR TITLE
[FIXED JENKINS-27436] <form> elements can’t be nested

### PR DESCRIPTION
**Pull request by ndeloof**
@reviewbybees

### CODE REVIEW

File: core/src/main/resources/hudson/PluginManager/check.jelly [hunk 1]
```diff
@@ -25,9 +25,9 @@ THE SOFTWARE.
 <j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
 
   <j:invokeStatic var="ds" className="jenkins.model.DownloadSettings" method="get"/>
-  <form method="post" action="${ds.useBrowser ? 'checkUpdates' : 'checkUpdatesServer'}">
-    ${%lastUpdated(app.updateCenter.lastUpdatedString)}
+  ${%lastUpdated(app.updateCenter.lastUpdatedString)}
+  <f:link href="${ds.useBrowser ? 'checkUpdates' : 'checkUpdatesServer'}" post="true">
     <f:submit value="${%Check now}" />
```
Comment (jglick) [inline core/src/main/resources/hudson/PluginManager/check.jelly]: An `<input type="submit">` inside an `<a href="…">`? I do not think so. Did you actually try this? The body of `<f:link>` is just supposed to be a label. The actual POSTing happens in `link.js` (`link.post`) via on `onclick` handler.

Comment (ndeloof) [inline core/src/main/resources/hudson/PluginManager/check.jelly]: I did try this and this worked fine.
I agree with you this looks bad, but can't find any other way to the link look like other buttons.

Comment (jglick) [inline core/src/main/resources/hudson/PluginManager/check.jelly]: Something is wrong here. The only purpose of `f:link` is to install an `onclick` handler (that dynamically creates and submits an invisible `form`), which you are not even using. So it sounds like your button is submitting some _other_ containing form, with an `action` unrelated to the `href` specified here.

Comment (ndeloof) [inline core/src/main/resources/hudson/PluginManager/check.jelly]: without this #@%! YUI stuff I'd just have used onclick="this.form.action='checkUpdates'". If you know which css class to use to get button look'n feel on a link, I'd be happy to remove this submit button.

**Comment by jtnord** [comment]
Note:  in HTML4 (not checked 5) it is illegal to have a form within a form.

**Comment by ndeloof** [comment]
@jtnord same in html5

**Comment by jtnord** [comment]
@ndeloof so this isn't fixing that the form taglib handles nested forms - but fixing some issue caused by the fact that the forms where nested?

File: core/src/main/resources/hudson/PluginManager/check.jelly (hunk 1 above)
Comment (jglick) [inline core/src/main/resources/hudson/PluginManager/check.jelly]: > If you know which css class to use to get button look'n feel on a link

Out of my depth I am afraid. In the meantime I would suggest

```xml
<f:link href="${ds.useBrowser ? 'checkUpdates' : 'checkUpdatesServer'}" post="true">${%Check now}</f:link>
```

as the supported and safe technique, even if the UI is not exactly what you want.

**Comment by ndeloof** [comment]
using a plain <button> with adequate css class is enough to have a reasonable looking button on bottom bar but structured as a link in html.

**Comment by jglick** [comment]
Scary but :+1:

**Comment by jglick** [comment]
(and yes anything that could be done to hasten the removal of YUI from Jenkins would be much appreciated)

**Comment by ndeloof** [comment]
YUI is so entangled with jenkins UI I don't think this is even possible. At least not within my 4 dimensions world, but some say there's actually 11 of them ... who knows

**Comment by daniel-beck** [comment]
Does this actually work in all currently supported browsers?

**Comment by tfennelly** [comment]
:+1: assuming it works

**Comment by tfennelly** [comment]
The version of YUI we are using it way out of date. I was trying to get build/assembly scripts for it from the guys behind the more recent versions. They told me to go away (politely).

**Comment by tfennelly** [comment]
I believe we could phase it out though.

**Pull request merged**